Lint tests

som-snytt opened the pull request:
Probing whether linting is worth the effort, and which lints are useful.

Tests should be clean, but really want structure: useful test names (description, link to ticket), bug vs feature coverage, simple type test (instead of assignment to unused variable) and similar, quarantine of deprecated APIs. For example, requiring parens and Unit for every test method is just boiler plate.

Test needs to be expressive when written, and then resistant to bit rot.

Time to check in with the minimum viable test rig.

som-snytt committed:
Lint Java in junit

sjrd commented on src/testkit/scala/tools/testkit/Unsymly.scala:
```diff
@@ -0,0 +1,36 @@
+ *  "Unseemly".
+ */
+object Unsymly {
+  def mksym: Symbol = macro SymlyImpl.impl
```
Is that really necessary? There are two tests using this, with a grand total of ~15 symbol creations. Can't we just spell out the constructors instead of relying on macros for this?

som-snytt replied:
Yeah, it was more of a lark. But thanks for noticing!

Similarly, I'd skip the specialized `->` support for just warning suppression.

som-snytt commented on test/junit/scala/collection/TraversableLikeTest.scala:
```diff
@@ -2,35 +2,16 @@ package scala.collection
-  class Baz[@specialized(Int) A]() extends IndexedSeq[A] {
-    def apply(i: Int) = ???
-    def length: Int = 0
-  }
```
Sample remains of previous edits of test; feature under test was removed.

som-snytt commented on test/junit/scala/collection/TraversableLikeTest.scala:
```diff
@@ -2,35 +2,16 @@ package scala.collection
-
   @Test
-  def test_SI10631 {
+  def `SI10631 calling seqview.map.last evaluates head twice`(): Unit = {
```
Having only the issue number means guessing what is being tested.

som-snytt commented on test/junit/scala/collection/TraversableOnceTest.scala:
```diff
@@ -2,102 +2,20 @@ package scala.collection
-    import Ordering.Double.IeeeOrdering
-    val seq = Seq(5.0, 3.0, Double.NaN, 4.0)
-    assert(seq.min.isNaN)
-    assert(seq.max.isNaN)
```
Moved to MinByMaxBy, duplication from strawman.

som-snytt commented on test/junit/scala/collection/WithFilterTest.scala:
```diff
@@ -13,7 +14,7 @@ class WithFilterTest {
         if x % 2 == 0
       } yield x + 1
-    val xs1t: List[Int] = xs1
+    @unused val xs1t: List[Int] = xs1
```
Common idiom for type test. Sometimes expressed as ascription `xs1: List[Int]` without assignment to var.

som-snytt commented on test/junit/scala/reflect/FieldAccessTest.scala:
```diff
@@ -1,14 +1,15 @@
-    def fn = () => x
+    locally {
+      () => x
+    }
```
This was bug 9306, it shouldn't matter where the function is written.

som-snytt commented on test/junit/scala/reflect/internal/InferTest.scala:
```diff
@@ -1,19 +1,14 @@
+import org.junit.Assert.assertTrue
+import org.junit.Test
+
+//import scala.tools.nsc.symtab.SymbolTableForUnitTesting
```
clean me

som-snytt commented on test/junit/scala/reflect/internal/InferTest.scala:
```diff
@@ -36,28 +31,28 @@ class InferTest extends BytecodeTesting {
 
       val tp0 = typeOf[D[A]]
-      assert(tp0.contains(asym))
+      assertTrue(tp0.contains(asym))
```
picking up the symboltable assert (from old refactoring) induces deprecation for one-arg form. Normally don't care assertTrue vs Predef.assert.

som-snytt commented on test/junit/scala/reflect/internal/LongNamesTest.scala:
```diff
@@ -37,7 +34,7 @@ class LongNamesTest {
     val testClass = compiler.classloader.loadClass("pkg.Test")
 
-    val output = testClass.newInstance().asInstanceOf[{ def test(): String }].test()
+    val output = testClass.getDeclaredConstructor().newInstance().asInstanceOf[{ def test(): String }].test()
```
Class.newInstance deprecated since JDK9.

som-snytt committed:
Lint junit tests

Don't rely on package privates in sys.process.

Syntax errors are not context-aware, so deprecated syntax
can't be silenced with deprecated annotation.

NthPortal commented on test/junit/scala/collection/immutable/LazyListLazinessTest.scala:
```diff
@@ -717,7 +715,7 @@ class LazyListLazinessTest {
         else {
           init.evaluate(i)
-          Some(i, i + 1)
+          Some((i, i + 1))
```
I am ashamed I made this mistake

som-snytt replied:
It's a feature not a ~bug~ mistake.

NthPortal reviewed (commented):
the stuff I'm familiar with looks good. I'll try to review the other stuff Soon™

som-snytt commented on test/junit/scala/reflect/internal/NamesTest.scala:
```diff
@@ -35,30 +35,30 @@ class NamesTest {
+    assertTrue(lookupTypeName("uhu".toCharArray) eq uy)
 
     assertThrows[AssertionError](lookupTypeName("dog".toCharArray), _ contains "not yet created")
     val d = newTermName("dog")
```
? Noticed it while adding a test.

som-snytt commented:
@SethTisue oh look,

> **This branch has no conflicts with the base branch**
> Merging can be performed automatically.

SethTisue commented on test/junit/scala/SerializationStabilityTest.scala:
```diff
@@ -100,9 +100,9 @@ object SerializationStability {
+    //val f0: Function0[Int] = () => 0
+    //val f1: Function1[Int, Int] = _ => 0
+    //val f2: Function2[Int, Int, Int] = (_, _) => 0
     
```
let's either remove entirely, or comment why it's commented-out

SethTisue commented on test/junit/scala/SerializationStabilityTest.scala:
```diff
@@ -100,9 +100,9 @@ object SerializationStability {
-    val f2: Function2[Int, Int, Int] = (_, _) => 0
+    //val f0: Function0[Int] = () => 0
+    //val f1: Function1[Int, Int] = _ => 0
+    //val f2: Function2[Int, Int, Int] = (_, _) => 0
```
let's either remove these lines entirely, or add a comment why they're commented-out

som-snytt replied:
I'll remove, but add f0-f2 to editor hot keys.

som-snytt replied:
oh I see, there were used in the following lines someone previously commented out.

SethTisue commented on build.sbt:
```diff
@@ -871,7 +871,11 @@ lazy val testkit = configureAsSubproject(project)
+  .settings(scalacOptions in Compile += "-Xlint:-adapted-args,-nullary-unit,_")
+  //.settings(scalacOptions in Compile += "-Xlint")
+  //.settings(scalacOptions in Compile += "-Werror")
+  .settings(javacOptions in Compile ++= Seq("-Xlint"))
```
let's either remove the commented-out lines entirely, or add a comment why they're commented-out

som-snytt replied:
What did the hash table hope to accomplish before it died? Everything in its bucket list.

`.aspirationalSettings(scalacOptions in Compile ++= Seq("-Werror", "-Xlint"))`

SethTisue commented on test/junit/scala/collection/immutable/SetTest.scala:
```diff
@@ -1,15 +1,14 @@
   @Test
   def test_SI8346_toSet_soundness(): Unit = {
-    val any2stringadd = "Disabled string conversions so as not to get confused!"
+    //val any2stringadd = "Disabled string conversions so as not to get confused!"
```
omit?

som-snytt replied:
I'll comment the import and omit this.

SethTisue commented on test/junit/scala/collection/mutable/ArrayBufferTest.scala:
```diff
@@ -146,18 +144,14 @@ class ArrayBufferTest {
     assertEquals(ArrayBuffer(0, 1), b3)
   }
 
-  @Test(expected = classOf[IndexOutOfBoundsException])
```
in this test and others in this file, doesn't the `@Test` annotation need to stay, just without the arguments?

SethTisue replied:
and if I'm right, I wonder if other files in the PR are affected as well

som-snytt replied:
Let me double-check, because I did make and correct that bad edit somewhere.

SethTisue commented on test/junit/scala/reflect/internal/util/WeakHashSetTest.scala:
```diff
@@ -1,6 +1,6 @@
 package scala.reflect.internal.util
 
-import org.junit.Assert._
+//import org.junit.Assert.{assertEquals, assertTrue}
```
omit?

som-snytt replied:
I'll comment `// TODO should use these`

SethTisue commented on test/junit/scala/tools/nsc/FileUtils.scala:
```diff
@@ -60,12 +60,14 @@ object FileUtils {
-    for (subDir <- allSubDirs.toList.sortBy(_.iterator().asScala.map(_.toString).toIterable)) {
+    // TODO
+    //for (subDir <- allSubDirs) {
+    for (subDir <- allSubDirs.toList.sortBy(_.iterator().asScala.map(_.toString).iterator.to(Iterable))) {
```
omit commented-out code?

som-snytt replied:
I'll comment something about the additional warning here, IIRC.

som-snytt replied:
I couldn't be bothered, the build output is the comment.

SethTisue reviewed (changes requested):
this is awesome 👏 

two categories of concern:
* omitted `@Test` annotations
* commented-out code

otherwise LGTM & ready for prompt merge before merge conflicts start piling up

som-snytt commented on test/junit/scala/collection/immutable/ArraySeqTest.scala:
```diff
@@ -80,8 +80,8 @@ class ArraySeqTest {
   }
-
-  @Test def checkSearch: Unit = SeqTests.checkSearch(ArraySeq(0 to 1000: _*), 15,  implicitly[Ordering[Int]])
+  
```
whitespace! or, greenspace!

som-snytt replied:
oh, that was the line that conflicted with another PR!

som-snytt commented on test/junit/scala/reflect/internal/ScopeTest.scala:
```diff
@@ -48,7 +42,7 @@ class ScopeTest {
     nested.unlink(bar)
-    assert(!nested.containsName(bar.name))
-    assert(outer.containsName(bar.name))
+    assertTrue(!nested.containsName(bar.name))
```
Missed assertFalse opportunity.

som-snytt committed:
Tweaks per review

som-snytt commented:
Twelve squared files changed, certified blunder-free by Seth's review and now quick re-review.

Oh, I guess a convo will not be marked outdated if the exact line of the comment is not edited, even if nearby lines are changed.
